### ISSUE TITLE
Unit Tests

### ISSUE THREAD

**Issue by DasCapschen** [Libre-TrainSim/Libre-TrainSim#255]
I think after the 0.9 release, we should try and move toward a more test-driven development cycle.

During the making of 0.9, a lot of refactors happened, and they often broke functionality, sometimes in subtle ways. I think many of such errors could have been caught with unit tests. 

Yes, it will be quite a bit of work to write all the unit tests at first, but once it is up and running, we should have a much smoother experience developing.

Possible Unit Testing addons
- https://github.com/bitwes/Gut
- :star: https://github.com/MikeSchulze/gdUnit3

**Comment by HaSa1002** [comment]
I am torn on unit tests for two reasons. First, we are a game. Meaning, we don't need to keep compatibility with users using our code in their projects, but only with user playing the game and thus have a different perception of bugs. On the other hand we do provide a platform for maps and mods. Second, they create overhead and increase the barrier of entry to contribute. However, they could provide some meaning to our CI runs and ensure that some of the known bugs don't happen again as long as they are caused by simple triggers and are not a result of complex inter-mechanic conjunctions.

TL;DR. I'd give a shot, because I am curious to find out how (well) a test driven development could work, but I would stick with our current approach to merge things quick and do necessary fixups later to bind the energy on and in this repository. More importantly I think, we should create test cases whenever we fix bugs.

Sill TL;DR. I am in favor of unit test, but have some concerns about how we implement the process.

**Comment by DasCapschen** [comment]
Unit tests are not about keeping compatibility, and they are not for users. They are only for us developers.
They are about not accidentally breaking something when you refactor something and introduce regressions.

I don't think the barrier to contribute is going to change much. We don't need to have 100% code coverage, I think that's impossible for a game.
And writing a test for even a complex function takes only minutes.

**Comment by HaSa1002** [comment]
Well IMO unit tests and refactors will never work together, because of the very nature of refactors you break stuff. But that is not the topic. I agree with the test coverage, but I am not so sure how easy tests are going to be as soon as we are working with scenes. There is probably only a relatively small subset of functions we can test. But anyhow refer to my answer above :D

**Comment by custodian** [comment]
My vote is for gdUnit3/4 if LTS migration happens along the way.
